File: NeetCode 150/sliding_window/Longest_repeating_character_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = r = m = 0
        res = 1

        hash_map = {}
        while (r <= len(s) - 1):
            hash_map[s[r]] = hash_map.get(s[r], 0) + 1
            m = max(m, hash_map[s[r]])
            while (not (r - l + 1 - m <= k)):
                hash_map[s[l]] = hash_map.get(s[l]) - 1
                l += 1
            res = max(res, r - l + 1)
            r += 1
        return res


sol = Solution()
res = sol.characterReplacement("ABAB", 2)
print(res)

File: NeetCode 150/sliding_window/LongestRepeatingCharacterReplacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */

//  a b b c b c c c

var characterReplacement = function (s, k) {
  let l = 0;
  let r = 0;
  let max = 0;
  let res = 1;
  const hashMap = new Map();

  while (r <= s.length - 1) {
    hashMap[s[r]] = hashMap[s[r]] + 1;
    console.log(`hashMap[${s[r]}] = ${hashMap[s[r]]}`);
    max = Math.max(...Object.values(hashMap));
    while (!(r - l + 1 - max <= k)) {
      hashMap[s[l]] = hashMap.get(s[l]) - 1;
      max = Math.max(...Object.values(hashMap));
      l++;
    }
    res = Math.max(res, r - l + 1);
    r++;
  }
  return res;
};

const res = characterReplacement("ABAB", 2);
console.log(res);
